nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/hive/streaming/NiFiRecordSerDe.java

```diff
@@ -0,0 +1,247 @@
+    protected TimestampParser tsParser;
+
+    private final static Pattern INTERNAL_PATTERN = Pattern.compile("_col([0-9]+)");
+
+    public NiFiRecordSerDe(RecordReader recordReader, ComponentLog log) {
+        this.recordReader = recordReader;
+        this.log = log;
+    }
+
+    @Override
+    public void initialize(Configuration conf, Properties tbl) {
+        List<TypeInfo> columnTypes;
+        StructTypeInfo rowTypeInfo;
+
+        log.debug("Initializing NiFiRecordSerDe: {}", tbl.entrySet().toArray());
+
+        // Get column names and types
+        String columnNameProperty = tbl.getProperty(serdeConstants.LIST_COLUMNS);
+        String columnTypeProperty = tbl.getProperty(serdeConstants.LIST_COLUMN_TYPES);
+        final String columnNameDelimiter = tbl.containsKey(serdeConstants.COLUMN_NAME_DELIMITER) ? tbl
+                .getProperty(serdeConstants.COLUMN_NAME_DELIMITER) : String.valueOf(SerDeUtils.COMMA);
+        // all table column names
+        if (columnNameProperty.isEmpty()) {
+            columnNames = new ArrayList<>(0);
+        } else {
+            columnNames = new ArrayList<>(Arrays.asList(columnNameProperty.split(columnNameDelimiter)));
+        }
+
+        // all column types
+        if (columnTypeProperty.isEmpty()) {
+            columnTypes = new ArrayList<>(0);
+        } else {
+            columnTypes = TypeInfoUtils.getTypeInfosFromTypeString(columnTypeProperty);
+        }
+
+        log.debug("columns: {}, {}", new Object[]{columnNameProperty, columnNames});
+        log.debug("types: {}, {} ", new Object[]{columnTypeProperty, columnTypes});
+
+        assert (columnNames.size() == columnTypes.size());
+
+        rowTypeInfo = (StructTypeInfo) TypeInfoFactory.getStructTypeInfo(columnNames, columnTypes);
+        schema = rowTypeInfo;
+        log.debug("schema : {}", new Object[]{schema});
+        cachedObjectInspector = (StandardStructObjectInspector) TypeInfoUtils.getStandardJavaObjectInspectorFromTypeInfo(rowTypeInfo);
+        tsParser = new TimestampParser(HiveStringUtils.splitAndUnEscape(tbl.getProperty(serdeConstants.TIMESTAMP_FORMATS)));
+    }
+
+    @Override
+    public Class<? extends Writable> getSerializedClass() {
+        return null;
+    }
+
+    @Override
+    public Writable serialize(Object o, ObjectInspector objectInspector) throws SerDeException {
+        return null;
+    }
+
+    @Override
+    public SerDeStats getSerDeStats() {
+        return null;
+    }
+
+    @Override
+    public Object deserialize(Writable writable) throws SerDeException {
+        ObjectWritable t = (ObjectWritable) writable;
+        Record record = (Record) t.get();
+        List<Object> r = new ArrayList<>(Collections.nCopies(columnNames.size(), null));
+        try {
+            RecordSchema recordSchema = record.getSchema();
+            for (RecordField field : recordSchema.getFields()) {
+                String fieldName = field.getFieldName();
```

Comment:
Can the field name order be different than table schema column names? Field re-ordering to match table schema is really expensive. May be we can have a fast path where if field order exactly matches as that of table schema we don't have to do all of these per field object extraction.

Reply:
We can require the incoming field names to be in the same order as the table schema column names, but historically we at least didn't care about the partition columns, they were removed from the object values as they had to be provided separately. If we require the partition columns be the last columns in the record, then we'll have to pay that expense somewhere, either upstream in NiFi or here. If we move the field name to position mapping to init, then we should only have to pay the price once per file and not per record?

Reply:
Yes. At the very least this should happen once per file and not per record. 
For unpartitioned or statically partitioned case, this should not happen as there is no partition columns to extract (static partitions goes into API) so should use fast path without these checks and allocations.

Reply:
For static partitions, the fields and their values will still be in the record. Are they ignored/skipped in that case?

Reply:
Ignored. Static partitions doesn't have to be in the record since its already specified in the API.

Reply:
True, I'm just saying in our Record API the partition columns will almost always be in the record. The generic use case is to use the PartitionRecord processor upstream, which is basically a GROUP BY, it takes one flow file in and sends out one for each unique partition value, so all the records in a flow file will have the same partition value (and it is added as metadata so you can use it later). In PutHive3Streaming you can specify the partition value as the value from the metadata, but the partition columns are still in the records.